Theme/Plugin Bundling: Handling upgrading to purchase a bundled theme

markbiek opened the pull request:
#### Proposed Changes

When purchasing a bundled theme from the showcase, the user needs to upgrade to a Business plan first.

After purchasing the Business plan, we redirect them to `designSetup` in stepper where they can activate the selected theme and finish the stepper Woo onboarding process.

#### Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->


1. Using a site on a Free/Personal/Premium plan
1. Go to `http://calypso.localhost:3000/themes/[site-slug]`
1. Pick the Thriving Artist theme
1. From the theme page, click "Upgrade to activate"
1. Go through the process of purchasing a Business plan
1. After checkout, you should be taken to Thriving Artist on `designSetup`
1. Click "Start with Thriving Artist"
1. You should be directed to the Woo stepper flow (starting with `storeAddress`)

#### Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?

<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to #

markbiek committed:
Handling upgrading to purchase a bundled theme

When purchasing a bundled theme from the showcase, the user needs to
upgrade to a Business plan first.

After purchasing the Business plan, we redirect them to `designSetup` in
stepper where they can activate the selected theme and finish the
stepper Woo onboarding process.

sirbrillig commented on client/my-sites/checkout/get-thank-you-page-url/index.ts:
```diff
@@ -135,6 +136,12 @@ export default function getThankYouPageUrl( {
+		// We need to do a hard redirect if we're redirecting to the stepper.
+		// Since stepper is self-contained, it doesn't load properly if we do a normal history state change
+		if ( isURL( redirectTo ) || redirectTo.startsWith( '/setup' ) ) {
+			window.location.href = redirectTo;
```
This function is covered by unit tests to prevent regressions. Could you add a unit test for this new behavior in `client/my-sites/checkout/get-thank-you-page-url/test/get-thank-you-page-url.ts`?

Also... 😬  is actually _performing_ a redirect in a function called `getThankYouPageUrl()` really a good idea? That seems like a very slippery slope for those who use this function (eg: it's called several times in checkout before a transaction is complete). Could this be moved to the code that actually performs the redirect instead?

sirbrillig reviewed (dismissed):
I think that the `getThankYouPageUrl()` function should not be made to perform redirects as it's used in many places that assume it will return a string.

markbiek commented:
> I think that the `getThankYouPageUrl()` function should not be made to perform redirects as it's used in many places that assume it will return a string.

That's a good point, thanks! I'll get that adjusted.

sirbrillig commented:
Check out `useRedirectOnTransactionSuccess` and its sub functions. I think that might be what you need.

sirbrillig commented:
See also `absoluteRedirectThroughPending` which can help to create the URL in the first place.

markbiek commented:
> `absoluteRedirectThroughPending`

Ah, good. I was just looking at that one. I could absolutely be mistaken but it seemed like using `absoluteRedirectThroughPending` in `useCreatePaymentCompleteCallback` (since that's where `getThankYouPageURL` gets called) might be more appropriate than changing `useRedirectOnTransactionSuccess`.

sirbrillig commented:
Sorry for the cryptic info. I'm AFK but I can help to find the right spot for this logic when I get back unless you locate it before then.

sirbrillig commented:
> Ah, good. I was just looking at that one. I could absolutely be mistaken but it seemed like using `absoluteRedirectThroughPending` in `useCreatePaymentCompleteCallback` (since that's where `getThankYouPageURL` gets called) might be more appropriate than changing `useRedirectOnTransactionSuccess`.

Yes, I think you're right. For example, [this spot](https://github.com/Automattic/wp-calypso/blob/bb44e8e7e40620436708e53ce43943e82e453810/client/my-sites/checkout/composite-checkout/hooks/use-create-payment-complete-callback.tsx#L229-L236) forces a full page load rather than an SPA redirect. As for where exactly to put this condition... you could try to add it [just before the regular redirect](https://github.com/Automattic/wp-calypso/blob/bb44e8e7e40620436708e53ce43943e82e453810/client/my-sites/checkout/composite-checkout/hooks/use-create-payment-complete-callback.tsx#L239), but if that's not sufficient we could try to force the URL to be absolute somewhere else (eg: in `getThankYouPageUrl` or maybe even in the `getUrl` function you're adding here. If you can use `absoluteRedirectThroughPending`, though, it will probably be the easiest).

markbiek commented:
@sirbrillig I think I'm confused about the intent behind `absoluteRedirectThrough Pending`.

At the point that I'm trying to call it in `useCreatePaymentCompleteCallback`, the value of `url` is the absolute url that I want to redirect to.

But `absoluteRedirectThroughPending` is taking my url and turning it into something like `http://calypso.localhost:3000/checkout/thank-you/no-site/pending/:orderId?redirectTo=<my url-encoded redirect url>`.

Is `absoluteRedirectThroughPending` really necessary? Could I not just do the redirect I need?

markbiek committed:
Move redirect out of get-thank-you-page-url

`getThankYouPageUrl` should just return a string, not perform any logic.
This moves our redirect code into `useCreatePaymentCompleteCallback`
(where `getThankYouPageUrl` is called).

We also leverage the `absoluteRedirectThroughPending` function to
properly complete checkout before redirecting.

spraveenitpro commented:
### Functional Testing:

1. I did a function testing of this until step 5, at step 6, this is what I saw:

![g579fG.png](https://user-images.githubusercontent.com/552016/193684091-8c470efb-511c-4d67-8b68-861db1f0ceac.png)

2. Now, I tried to get to the designSetup by clicking the below which instead launched the FSE:

![Fsum8l.png](https://user-images.githubusercontent.com/552016/193684006-baef39ee-826b-4378-a779-a8191facf892.png)

3. If I am missing something then let me know.

sirbrillig commented:
> I did a function testing of this until step 5, at step 6, this is what I saw

@spraveenitpro `window.appBoot is not a function` usually happens when you try to switch calypso branches without re-starting calypso. Maybe try stopping it and starting it up again?

> At the point that I'm trying to call it in useCreatePaymentCompleteCallback, the value of url is the absolute url that I want to redirect to. But absoluteRedirectThroughPending is taking my url and turning it into something like http://calypso.localhost:3000/checkout/thank-you/no-site/pending/:orderId?redirectTo=<my url-encoded redirect url>. Is absoluteRedirectThroughPending really necessary? Could I not just do the redirect I need?

@markbiek  Sorry, I should have explained all this in more detail earlier! Let me back up.

All post-checkout URLs must go through the "pending" page for several reasons (unified analytics, provisioning, and notifications for both direct and redirect-based payment methods). Fortunately, most visits to the page are extremely brief and users may not see it render at all. Its user-facing purpose (when a payment is actually complete anyway) is just to redirect to the actual post-checkout page returned by `getThankYouPageUrl()`. 

When checkout is completed with a non-redirect payment method, the callback created by `useCreatePaymentCompleteCallback()` will be called. It will create a pending page URL by getting the final URL from `getThankYouPageUrl()` and then passing it to `redirectThroughPending()` or `absoluteRedirectThroughPending()`. Then it will redirect to that (pending) URL which will in turn redirect the user to the final URL. 

If the URL is not absolute, this final redirect [will be performed](https://github.com/Automattic/wp-calypso/blob/bb44e8e7e40620436708e53ce43943e82e453810/client/my-sites/checkout/checkout-thank-you/pending/index.tsx#L116-L122) as a Single Page App redirect using `page()`. If the final URL is absolute, this final redirect will be performed with `window.location`.

(For redirect payment methods, the browser will go directly from the 3rd-party site to the pending page, and the above steps will proceed from there.)

So if your goal is to make sure that we use `window.location` in the final redirect when the URL meets certain conditions, in theory all we have to do is make sure the URL is absolute (in this case defined as a URL that does not start with `/`).

If, as you say, the URL you're using (the final URL returned by `getThankYouPageUrl()`) is already absolute and you're not getting a full-page redirect, then something must be wrong with the above flow. I'll try out this branch to see if I can find out what's going on.

sirbrillig commented on client/my-sites/checkout/composite-checkout/hooks/use-create-payment-complete-callback.tsx:
```diff
@@ -237,6 +238,13 @@ export default function useCreatePaymentCompleteCallback( {
+			// We need to do a hard redirect if we're redirecting to the stepper.
+			// Since stepper is self-contained, it doesn't load properly if we do a normal history state change
+			if ( isURL( url ) || url.startsWith( '/setup' ) ) {
+				window.location.href = url;
```
We must use the pending page, but I think that if you change this to the following, it will be fine:

```
				absoluteRedirectThroughPending( url, {
					siteSlug,
					orderId: transactionResult.order_id,
					receiptId: transactionResult.receipt_id,
				} );
```

However, you might actually not need this block at all if you make the other change I mentioned below.

markbiek replied:
Ah, I see what you mean. I think changing to an absolute url makes the most sense. I'll try that.

sirbrillig commented on client/my-sites/themes/theme-options.js:
```diff
@@ -86,7 +86,14 @@ function getAllThemeOptions( { translate, isFSEActive } ) {
+		getUrl: ( state, themeId, siteId ) => {
+			const slug = getSiteSlug( state, siteId );
+			const redirectTo = encodeURIComponent(
+				`/setup/designSetup?siteSlug=${ slug }&theme=${ themeId }`
```
Since this is a relative URL, the redirect from the pending page to `/setup/designSetup` will always use a Single Page App redirect. If you want this to get a full page reload, just change this URL to be absolute instead. (It must start with `https`.) Of course, constructing an absolute URL is not trivial since the user could be on wordpress.com, wpcalypso.wordpress.com, calypso.localhost, etc. You can use a technique [like this](https://github.com/Automattic/wp-calypso/blob/f8cb8cd8f0df2f8131e3c2b5eb20e31f0cfc9b7a/client/my-sites/checkout/composite-checkout/lib/pending-page.ts#L164) to figure it out.

sirbrillig reviewed (commented):
I think I figured it out. The redirect URL was relative, which is why it was using a non-absolute redirect.

daledupreez commented on client/landing/stepper/declarative-flow/internals/steps-repository/design-setup/unified-design-picker.tsx:
```diff
@@ -153,6 +154,25 @@ const UnifiedDesignPickerStep: Step = ( { navigation, flow } ) => {
+
+		const { generated: generatedDesigns, static: staticDesigns } = allDesigns;
+
+		const designs = [ ...generatedDesigns.designs, ...staticDesigns.designs ];
+
+		for ( const design of designs ) {
+			if ( design.slug === themeFromQueryString ) {
+				setSelectedDesign( design );
+				setIsPreviewingDesign( true );
+			}
+		}
```
Some small notes on this implementation:
1. We're not breaking out of the loop when we find the expected design.
2. The for loop contains both our search logic and our resulting actions. (This one is way more of a preference.)

It may be easier to split this into separate steps using `find()` and an early exit condition.
```suggestion
		const designs = [ ...generatedDesigns.designs, ...staticDesigns.designs ];

		const requestedDesign = designs.find( ( design ) =>  design.slug === themeFromQueryString );
		if ( ! requestedDesign ) {
			return;
		}
		
		setSelectedDesign( requestedDesign );
		setIsPreviewingDesign( true );
```

markbiek replied:
Much tidier, thanks!

daledupreez reviewed (commented):
I've dropped a small suggestion, but I've updated the PR labels to reflect that we still need to incorporate some of sirbrillig's feedback.

markbiek commented:
@spraveenitpro That `window.AppBoot is not a function` error is actually a known issue with the Stepper framework and not related to this PR.

Often times when you make a code change in Stepper (or change to a branch with Stepper changes) while Calypso is running, the build doesn't happen correctly leading to the above error.

Also, in this context, `designSetup` refers to the `/setup/designSetup` step in the onboarding Stepper flow rather than part of the FSE.

markbiek commented:
@sirbrillig @daledupreez I think this is ready for re-review!

sirbrillig reviewed (commented):
I tried to run through the testing instructions, but I couldn't find a "Thriving Artist" theme (I tried an old site without a plan, a brand new site, and even looked in the logged-out theme showcase). I thought that maybe the instructions would work for any premium theme, but I don't see an `Upgrade to activate` button anywhere on the Premium theme pages that I tried, so I must be doing something wrong. 

I won't mark it approved since I wasn't able to test, but in any case, the code changes here look fine to me from a Payments perspective so if your team approves it, then I have no objections.

markbiek commented:
> I tried to run through the testing instructions, but I couldn't find a "Thriving Artist" theme (I tried an old site without a plan, a brand new site, and even looked in the logged-out theme showcase). I thought that maybe the instructions would work for any premium theme, but I don't see an `Upgrade to activate` button anywhere on the Premium theme pages that I tried, so I must be doing something wrong.
> 
> I won't mark it approved since I wasn't able to test, but in any case, the code changes here look fine to me from a Payments perspective so if your team approves it, then I have no objections.

Thriving Artist is currently soft-launched so you'll need to be logged in with an A8C account. This change only applies to themes with bundled plugins (which is only Thriving Artist at the moment).

At any rate, thanks a ton for the help!

nelsonec87 reviewed (approved):
LGTM!
I've tested it, and it worked as described!

valterlorran reviewed (approved):
LGTM!
